duckdb-vortex/src/vortex_scan.cpp

```diff
@@ -211,77 +230,122 @@ static unique_ptr<VortexFileReader> OpenFileAndVerify(FileSystem &fs, const std:
+//
+// Returns true if a new array was assigned. Returns false otherwise.
+static bool GetNextArray(ClientContext &context, const VortexBindData &bind_data, VortexScanGlobalState &global_state,
+                         VortexScanLocalState &local_state, DataChunk &output) {
+
+	auto partition_idx = global_state.next_partition.fetch_add(1);
+
+	// No more partitions to read.
+	if (partition_idx >= global_state.scan_partitions.size()) {
+		global_state.finished = true;
+		return false;
+	}
 
 	if (local_state.array == nullptr) {
-		auto &slot = global_state.file_slots[local_state.thread_id];
-		std::lock_guard _l(slot.slot_lock);
+		auto partition = global_state.scan_partitions[partition_idx];
+
+		std::shared_ptr<VortexLayoutReader> layout_reader = [&] {
+			std::lock_guard<std::mutex> lock(global_state.layout_mutexes[partition.file_idx]);
+
+			if (global_state.layout_readers[partition.file_idx]) {
+				return global_state.layout_readers[partition.file_idx];
+			} else {
+				auto file_name = global_state.expanded_files[partition.file_idx];
+				auto vortex_file = OpenFileAndVerify(FileSystem::GetFileSystem(context), file_name, bind_data);
+				global_state.layout_readers[partition.file_idx] = VortexLayoutReader::CreateFromFile(vortex_file.get());
+				return global_state.layout_readers[partition.file_idx];
+			}
+		}();
 
-		if (global_state.finished.load() && slot.array_stream == nullptr) {
-			output.SetCardinality(0);
-			return;
-		}
+		local_state.stream = OpenArrayStream(global_state, layout_reader, partition);
+	}
 
-		// 1. check we can make progress on current owned file
-		// 2. check we can get another file
-		// todo: 3. check if we can work steal from another thread
-		// 4. we are done
+	local_state.array = local_state.stream->NextArray();
 
-		while (local_state.array == nullptr) {
-			if (slot.array_stream == nullptr) {
-				auto file_idx = global_state.next_file.fetch_add(1);
+	// Reset row offset for the array.
+	local_state.current_row = 0;
 
-				if (file_idx >= global_state.expanded_files.size()) {
-					local_state.finished = true;
-					global_state.finished = true;
-					output.Reset();
-					return;
-				}
+	// If the stream is empty, mark it as read by returning false.
+	if (local_state.array == nullptr) {
+		local_state.stream = nullptr;
+		return false;
+	}
 
-				auto file_name = global_state.expanded_files[file_idx];
-				auto file = OpenFileAndVerify(FileSystem::GetFileSystem(context), file_name, bind_data);
+	return true;
+}
 
-				slot.array_stream = OpenArrayStream(bind_data, global_state, file.get());
-			}
+static void VortexScanFunction(ClientContext &context, TableFunctionInput &data, DataChunk &output) {
+	auto &bind_data = data.bind_data->Cast<VortexBindData>();
+	auto &global_state = data.global_state->Cast<VortexScanGlobalState>();
+	auto &local_state = data.local_state->Cast<VortexScanLocalState>();
+
+	if (global_state.finished && local_state.stream == nullptr) {
```

Comment:
Swap the order of local and global state access it will be slightly faster in the case local_state is non null